src/arm_linux.rs

```diff
@@ -54,13 +56,43 @@ fn insert_aligned(aligned: u32, val: u32, shift: u32, mask: u32) -> u32 {
     (aligned & !(mask << shift)) | ((val & mask) << shift)
 }
 
+/// Loads the value at `ptr`.
+///
+/// # Safety
+///
+/// `ptr` must be aligned and point to memory within a page that allows read access.
+/// If `T` has a size of 4, then `ptr` must be valid for an atomic read.
+unsafe fn load_aligned<T>(ptr: *mut u32) -> u32 {
+    if mem::size_of::<T>() == 4 {
+        // SAFETY: As `T` has a size of 4, the caller garantees this is sound.
+        unsafe { AtomicU32::from_ptr(ptr).load(Ordering::Relaxed) }
+    } else {
+        // SAFETY:
+        // As all 4 bytes pointed to by `ptr` might not be dereferenceable due to the provenance of
+        // `ptr` when doing atomic operations on a `u8`/`i8`/`u16`/`i16`, inline ASM is used to
```

Comment:
It's not just about provenance. it might be genuinely outside the bounds of the allocation.